go/signedexchange/internal/signingalgorithm/signingalgorithm.go

```diff
@@ -43,8 +47,47 @@ func SigningAlgorithmForPrivateKey(pk crypto.PrivateKey, rand io.Reader) (Signin
 		case elliptic.P384().Params().Name:
 			return &ecdsaSigningAlgorithm{pk, crypto.SHA384, rand}, nil
 		default:
-			return nil, fmt.Errorf("signedexchange: unknown ECDSA curve: %s", name)
+			return nil, fmt.Errorf("signingalgorithm: unknown ECDSA curve: %s", name)
+		}
+	}
+	return nil, fmt.Errorf("signingalgorithm: unknown private key type: %T", pk)
+}
+
+type Verifier interface {
+	Verify(msg, sig []byte) (bool, error)
+}
+
+type ecdsaVerifier struct {
+	pubKey *ecdsa.PublicKey
+	hash   crypto.Hash
+}
+
+func (e *ecdsaVerifier) Verify(msg, sig []byte) (bool, error) {
+	var v ecdsaSigValue
+	rest, err := asn1.Unmarshal(sig, &v)
+	if err != nil {
+		return false, fmt.Errorf("signingalgorithm: failed to ASN.1 decode the signature: %v", err)
+	}
+	if len(rest) > 0 {
+		return false, errors.New("signingalgorithm: extra data at the signature end")
+	}
+
+	hash := e.hash.New()
+	hash.Write(msg)
+	return ecdsa.Verify(e.pubKey, hash.Sum(nil), v.R, v.S), nil
+}
+
+func VerifierForPublicKey(k crypto.PublicKey) (Verifier, error) {
```

Comment:
Do we plan to add another verifier other than `ecdsaVerifier`? If not, I think exporting `ecdsaVerifier` and returning it instead of returning an interface value should be fine.

Reply:
We may support ed25519 (and/or other non-legacy, non-RSA signing algorithms defined by TLS 1.3 or later) in the future.

I'd like to use this interface in the first place, to minimize caller-side changes when we support them.

Reply:
Makes sense.